### PR TITLE
Module 3 , Block 3 : Bookstore: add reducers and actions

### PR DESCRIPTION
**Pull request by aceDavon**
## In this PR, I have implemented the following tasks: :writing_hand: 

- [ ] Use store, actions, and reducers in React
- [ ] Add Redux Toolkit
- [ ] Structure application files using a "feature folder" approach and the ducks pattern
- [ ] Write book's actions and reducer
- [ ] Write categories' actions and reducer
- [ ] Configure the Redux Store

### CODE REVIEW

File: src/redux/books/BookSlice.js [hunk 1]
```diff
@@ -0,0 +1,15 @@
+export const addBook = 'ADD_BOOK';
+export const removeBook = 'REMOVE_BOOK';
```
Comment (ShoiraTa) [inline src/redux/books/BookSlice.js]: - [ ] The value of each action type needs to follow the [ducks pattern](https://github.com/erikras/ducks-modular-redux), so it will be something like:       
`const ADD_BOOK = bookstore/books/ADD_BOOK` and 
`const REMOVE_BOOK = bookstore/books/REMOVE_BOOK`

File: src/redux/categories/CategorySlice.js [hunk 2]
```diff
@@ -0,0 +1,12 @@
+export const checkStatus = 'CHECK_STATUS';
```
Comment (ShoiraTa) [inline src/redux/categories/CategorySlice.js]: - [ ] Same here, kindly follow the [ducks pattern](https://github.com/erikras/ducks-modular-redux)

File: src/redux/books/BookSlice.js [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+
+const booksReducer = (state = [], action) => {
+  switch (action.type) {
+    case addBook:
+      return state.concat(action.payload);
```
Comment (ShoiraTa) [inline src/redux/books/BookSlice.js]: - [ ] Here you are mutating the state, but mutating state is an anti-pattern in React. React uses a rendering engine which depends on the fact that state changes are observable. This observation is made by comparing previous state with next state. It will alter a virtual dom with the differences and write changed elements back to the dom.

When you alter the internal state, React does not know what's changed, and even worse; it's notion of the current state is incorrect. So the dom and virtual dom will become out of sync.

Redux uses the same idea to update it's store; an action can be observed by reducers which calculate the next state of the store. Changes are emitted and for example consumed by react-redux connect.

So in short: never ever, mutate state. Instead of Object.assign you can use the stage-3 spread syntax:

{...previousState,...changes} 

In your case it would be something like:

```
    case ADD_BOOK:
    { return [...state, action.payload]; }
```

File: src/redux/books/BookSlice.js [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+    case removeBook:
+      return state.map((book) => book.id !== action.payload);
```
Comment (ShoiraTa) [inline src/redux/books/BookSlice.js]: - [ ] Kindly use `filter` instead of map:
`return state.filter((book) => book.d !== action.payload);`

**Review by ShoiraTa — CHANGES_REQUESTED** [review]
Hi @aceDavon,

:thumbsup: This is good to make it this far, your project is really nice, congratulations on the good work you have done so far.

**First let me to point out:**

- Great commit history! ✔️ 
- You passed all the Linters! ✔️ 
- Professional pull request! ✔️ 

Keep on coding, good job so far! :clap:

There are some issues that you still need to work on to go to the next project but you are almost there!



 ## Required Changes  :recycle: 

_Check the comments under the review._ 

***
Cheers and Happy coding! :clap: :clap: :clap:

Feel free to leave any questions or comments in the PR thread if something is not 100% clear by mentioning my Github username.


**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

* * *
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Review by mukhammadrizooff — APPROVED** [review]
Hi @aceDavon, 

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



## Highlights
- No linter errors ✔️ 
- Good implementations ✔️ 

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._
